src/machines/java/com/enderio/machines/common/blockentity/VatBlockEntity.java

```diff
@@ -0,0 +1,260 @@
+            }
+        }
+        return super.onBlockEntityUsed(state, level, pos, player, hand, hit);
+    }
+
+    protected VatBlockEntity.VatCraftingMachineTask createTask(Level level, FermentingRecipe.Input input,
+        @Nullable RecipeHolder<FermentingRecipe> recipe) {
+        return new VatBlockEntity.VatCraftingMachineTask(level, getInventoryNN(), getFluidHandler(), input, recipe);
+    }
+
+    @Override
+    public @Nullable MachineInventoryLayout getInventoryLayout() {
+        return MachineInventoryLayout.builder().inputSlot(2).slotAccess(REAGENTS).build();
+    }
+
+    @Override
+    protected void onInventoryContentsChanged(int slot) {
+        super.onInventoryContentsChanged(slot);
+        craftingTaskHost.newTaskAvailable();
+    }
+
+    private FermentingRecipe.Input createRecipeInput() {
+        List<ItemStack> reagents = REAGENTS.getItemStacks(getInventoryNN());
+        return new FermentingRecipe.Input(reagents.get(0), reagents.get(1), getInputTank());
+    }
+
+    @Override
+    public MachineTankLayout getTankLayout() {
+        return MachineTankLayout
+            .builder()
+            .tank(INPUT_TANK, TANK_CAPACITY, true, false, (stack) -> true)
+            .tank(OUTPUT_TANK, TANK_CAPACITY, false, true, (stack) -> true)
+            .build();
+    }
+
+    @Override
+    public MachineFluidHandler getFluidHandler() {
+        return fluidHandler;
+    }
+
+    @Override
+    public MachineFluidHandler createFluidHandler() {
+        return new MachineFluidHandler(this, getTankLayout()) {
+            @Override
+            protected void onContentsChanged(int slot) {
+                super.onContentsChanged(slot);
+                craftingTaskHost.newTaskAvailable();
+                setChanged();
+            }
+        };
+    }
+
+    public MachineFluidTank getInputTank() {
+        return INPUT_TANK.getTank(this);
+    }
+
+    public MachineFluidTank getOutputTank() {
+        return OUTPUT_TANK.getTank(this);
+    }
+
+    public float getProgress() {
+        return craftingTaskHost.getProgress();
+    }
+
+    public ResourceLocation getRecipeId() {
+        if (level.isClientSide()) {
+            return recipeId;
+        }
+
+        if (craftingTaskHost.getCurrentTask() != null) {
+            ResourceLocation id = craftingTaskHost.getCurrentTask().getRecipeId();
+            return id != null ? id : EMPTY;
+        }
+        return EMPTY;
+    }
+
+    public void setRecipeId(ResourceLocation recipeId) {
+        this.recipeId = recipeId;
+    }
+
+    public CraftingMachineTaskHost<FermentingRecipe, FermentingRecipe.Input> getCraftingHost() {
+        return craftingTaskHost;
+    }
+
+    public void moveFluidToOutputTank() {
+        if (level != null && level.isClientSide()) {
+            PacketDistributor.sendToServer(new VatMoveTankPacket(getBlockPos()));
+        } else {
+            if (OUTPUT_TANK.isEmpty(this) && !INPUT_TANK.isEmpty(this)) {
```

Comment:
instead of empty, it should check if it can be filled. As a side note, why do we have this? What use does it have?

Reply:
isEmpty was done to ensure that fluid matches and dumped in entirety. Take hootch recipe, it converts water into hootch. so to transfer water over, I am simply checking if tank is empty and copying over fluidstacks instead of doing a transfer test

Reply:
Yes but imagine you have both hootch in the input and output. Then you can't move the input to the output and that's confusing. Also still don't know why this is a feature. Wouldn't output to input be more usefull?

Reply:
This was how it was in 1.12 so I replicated it 1:1. I think it is because the input thank is input only so there is no way get rid of tank contents. and output tank is output only so that transfer button can move fluids between tanks.